File: Tutorials/Codes/Autonomous_Mobile_Robots_Course/Chapter15/Lesson3/Chapter15_Lesson3.java
/*
Chapter15_Lesson3.java
Dynamic Window Approach (DWA) — minimal Java implementation (no ROS required).
Compile:
  javac Chapter15_Lesson3.java
Run:
  java Chapter15_Lesson3
*/

import java.util.ArrayList;
import java.util.List;

public class Chapter15_Lesson3 {

    static class State {
        double x, y, theta, v, w;
        State(double x, double y, double theta, double v, double w) {
            this.x = x; this.y = y; this.theta = theta; this.v = v; this.w = w;
        }
    }

    static class Config {
        double vMin = -0.2, vMax = 1.0;
        double wMin = -1.5, wMax = 1.5;

        double aMax = 0.8;
        double aBrake = 1.0;
        double alphaMax = 2.0;

        double dt = 0.1;
        double T = 2.0;
        double vRes = 0.05;
        double wRes = 0.1;

        double robotRadius = 0.3;

        double wHeading = 0.4;
        double wClear = 0.4;
        double wSpeed = 0.2;
    }

    static class Candidate {
        double h, c, s, v, w;
        List<State> traj;
        Candidate(double h, double c, double s, double v, double w, List<State> traj) {
            this.h = h; this.c = c; this.s = s; this.v = v; this.w = w; this.traj = traj;
        }
    }

    static double wrapToPi(double a) {
        double pi = Math.PI;
        a = (a + pi) % (2.0 * pi);
        if (a < 0) a += 2.0 * pi;
        return a - pi;
    }

    static State motionStep(State s, double v, double w, double dt) {
        double nx = s.x + v * Math.cos(s.theta) * dt;
        double ny = s.y + v * Math.sin(s.theta) * dt;
        double nt = s.theta + w * dt;
        return new State(nx, ny, nt, v, w);
    }

    static double[] dynamicWindow(State s, Config cfg) {
        double vLowDyn = s.v - cfg.aBrake * cfg.dt;
        double vHighDyn = s.v + cfg.aMax * cfg.dt;
        double wLowDyn = s.w - cfg.alphaMax * cfg.dt;
        double wHighDyn = s.w + cfg.alphaMax * cfg.dt;

        double vL = Math.max(cfg.vMin, vLowDyn);
        double vU = Math.min(cfg.vMax, vHighDyn);
        double wL = Math.max(cfg.wMin, wLowDyn);
        double wU = Math.min(cfg.wMax, wHighDyn);
        return new double[]{vL, vU, wL, wU};
    }

    static List<State> rollout(State s0, double v, double w, Config cfg) {
        int n = (int)Math.floor(cfg.T / cfg.dt);
        List<State> traj = new ArrayList<>();
        traj.add(s0);
        State s = s0;
        for (int i = 0; i < n; i++) {
            s = motionStep(s, v, w, cfg.dt);
            traj.add(s);
        }
        return traj;
    }

    static double minClearance(List<State> traj, double[][] obs) {
        double best = Double.POSITIVE_INFINITY;
        for (State st : traj) {
            for (double[] o : obs) {
                double dx = st.x - o[0];
                double dy = st.y - o[1];
                double d = Math.hypot(dx, dy);
                if (d < best) best = d;
            }
        }
        return best;
    }

    static double headingScore(List<State> traj, double gx, double gy) {
        State last = traj.get(traj.size() - 1);
        double dir = Math.atan2(gy - last.y, gx - last.x);
        double err = wrapToPi(dir - last.theta);
        return Math.cos(err);
    }

    static double speedScore(double v, Config cfg) {
        return (v - cfg.vMin) / Math.max(1e-9, (cfg.vMax - cfg.vMin));
    }

    static boolean admissible(double v, double clear, Config cfg) {
        double vv = Math.max(0.0, v);
        double dStop = (vv * vv) / Math.max(1e-9, 2.0 * cfg.aBrake);
        return clear > (dStop + cfg.robotRadius);
    }

    static double[] normalize(double[] vals) {
        double lo = Double.POSITIVE_INFINITY, hi = Double.NEGATIVE_INFINITY;
        for (double v : vals) { lo = Math.min(lo, v); hi = Math.max(hi, v); }
        double[] out = new double[vals.length];
        if (Math.abs(hi - lo) < 1e-12) return out; // all zeros
        for (int i = 0; i < vals.length; i++) out[i] = (vals[i] - lo) / (hi - lo);
        return out;
    }

    static Object[] dwaControl(State s, double gx, double gy, double[][] obs, Config cfg) {
        double[] win = dynamicWindow(s, cfg);
        double vL = win[0], vU = win[1], wL = win[2], wU = win[3];

        List<Candidate> cand = new ArrayList<>();
        for (double v = vL; v <= vU + 1e-12; v += cfg.vRes) {
            for (double w = wL; w <= wU + 1e-12; w += cfg.wRes) {
                List<State> traj = rollout(s, v, w, cfg);
                double clear = minClearance(traj, obs) - cfg.robotRadius;
                if (!admissible(v, clear, cfg)) continue;
                double h = headingScore(traj, gx, gy);
                double sp = speedScore(v, cfg);
                cand.add(new Candidate(h, clear, sp, v, w, traj));
            }
        }

        if (cand.isEmpty()) return new Object[]{0.0, 0.0, List.of(s)};

        double[] hs = new double[cand.size()];
        double[] cs = new double[cand.size()];
        double[] ss = new double<cand.size()];
        for (int i = 0; i < cand.size(); i++) {
            hs[i] = cand.get(i).h;
            cs[i] = cand.get(i).c;
            ss[i] = cand.get(i).s;
        }
        hs = normalize(hs);
        cs = normalize(cs);
        ss = normalize(ss);

        double bestJ = -1e18;
        int bestIdx = 0;
        for (int i = 0; i < cand.size(); i++) {
            double J = cfg.wHeading * hs[i] + cfg.wClear * cs[i] + cfg.wSpeed * ss[i];
            if (J > bestJ) { bestJ = J; bestIdx = i; }
        }
        Candidate best = cand.get(bestIdx);
        return new Object[]{best.v, best.w, best.traj};
    }

    public static void main(String[] args) {
        Config cfg = new Config();
        State s = new State(0.0, 0.0, 0.0, 0.0, 0.0);
        double gx = 8.0, gy = 6.0;

        List<double[]> obsList = new ArrayList<>();
        for (int i = 0; i < 15; i++) obsList.add(new double[]{2.0 + i*(5.0/14.0), 3.0});
        for (int i = 0; i < 12; i++) obsList.add(new double[]{4.0, 1.0 + i*(4.0/11.0)});
        double[][] obs = obsList.toArray(new double[0][0]);

        for (int k = 0; k < 600; k++) {
            Object[] out = dwaControl(s, gx, gy, obs, cfg);
            double v = (double)out[0];
            double w = (double)out[1];
            s = motionStep(s, v, w, cfg.dt);
            double dist = Math.hypot(s.x - gx, s.y - gy);
            if (k % 20 == 0) {
                System.out.printf("k=%d x=%.3f y=%.3f dist=%.3f v=%.2f w=%.2f%n", k, s.x, s.y, dist, v, w);
            }
            if (dist < 0.3) break;
        }
        System.out.printf("Final: x=%.3f y=%.3f theta=%.3f%n", s.x, s.y, s.theta);
    }
}

File: Tutorials/Codes/Autonomous_Mobile_Robots_Course/Chapter15/Lesson3/Chapter15_Lesson3.py
"""
Chapter15_Lesson3.py
Dynamic Window Approach (DWA) — from-scratch reference implementation (2D unicycle).
- No ROS dependencies. Uses obstacle points and a simple receding-horizon rollout.
- You can later wrap `dwa_control()` inside a ROS2 node that publishes geometry_msgs/Twist.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Tuple
import math
import numpy as np


@dataclass
class State:
    x: float
    y: float
    theta: float   # heading [rad]
    v: float       # linear velocity [m/s]
    w: float       # angular velocity [rad/s]


@dataclass
class Config:
    # Kinematic limits
    v_min: float = -0.2
    v_max: float =  1.0
    w_min: float = -1.5
    w_max: float =  1.5

    # Dynamic limits (accelerations)
    a_max: float = 0.8          # linear accel [m/s^2]
    a_brake: float = 1.0        # linear decel magnitude [m/s^2]
    alpha_max: float = 2.0      # angular accel [rad/s^2]

    # Discretization
    dt: float = 0.1             # control period [s]
    T: float = 2.0              # rollout horizon [s]
    v_res: float = 0.05         # sampling resolution [m/s]
    w_res: float = 0.1          # sampling resolution [rad/s]

    # Robot geometry
    robot_radius: float = 0.3   # [m]

    # Scoring weights
    w_heading: float = 0.4
    w_clear: float = 0.4
    w_speed: float = 0.2


def motion_step(s: State, v: float, w: float, dt: float) -> State:
    """Forward Euler unicycle integration."""
    nx = s.x + v * math.cos(s.theta) * dt
    ny = s.y + v * math.sin(s.theta) * dt
    ntheta = s.theta + w * dt
    return State(nx, ny, ntheta, v, w)


def dynamic_window(s: State, cfg: Config) -> Tuple[float, float, float, float]:
    """
    Dynamic window:
      - kinematic window: [v_min, v_max] x [w_min, w_max]
      - acceleration window: [v - a_brake*dt, v + a_max*dt] x [w - alpha*dt, w + alpha*dt]
      - intersection
    """
    v_low_kin, v_high_kin = cfg.v_min, cfg.v_max
    w_low_kin, w_high_kin = cfg.w_min, cfg.w_max

    v_low_dyn = s.v - cfg.a_brake * cfg.dt
    v_high_dyn = s.v + cfg.a_max * cfg.dt
    w_low_dyn = s.w - cfg.alpha_max * cfg.dt
    w_high_dyn = s.w + cfg.alpha_max * cfg.dt

    v_low = max(v_low_kin, v_low_dyn)
    v_high = min(v_high_kin, v_high_dyn)
    w_low = max(w_low_kin, w_low_dyn)
    w_high = min(w_high_kin, w_high_dyn)
    return v_low, v_high, w_low, w_high


def rollout(s0: State, v: float, w: float, cfg: Config) -> List[State]:
    n = int(cfg.T / cfg.dt)
    traj = [s0]
    s = s0
    for _ in range(n):
        s = motion_step(s, v, w, cfg.dt)
        traj.append(s)
    return traj


def min_clearance(traj: List[State], obstacles: np.ndarray) -> float:
    """
    Obstacles: array shape (M,2) of point obstacles in world coordinates.
    Returns minimum Euclidean distance from any trajectory point to any obstacle.
    """
    pts = np.array([(st.x, st.y) for st in traj], dtype=float)  # (N,2)
    # pairwise distances: (N,M)
    d = np.linalg.norm(pts[:, None, :] - obstacles[None, :, :], axis=2)
    return float(np.min(d))


def heading_score(traj: List[State], goal: Tuple[float, float]) -> float:
    """
    Heading score: cosine alignment between final heading and direction to goal.
    Range roughly [-1, 1], higher is better.
    """
    gx, gy = goal
    last = traj[-1]
    dir_to_goal = math.atan2(gy - last.y, gx - last.x)
    ang_err = wrap_to_pi(dir_to_goal - last.theta)
    return math.cos(ang_err)


def speed_score(v: float, cfg: Config) -> float:
    """Prefer higher forward speed within limits."""
    return (v - cfg.v_min) / max(1e-9, (cfg.v_max - cfg.v_min))


def admissible(v: float, clearance: float, cfg: Config) -> bool:
    """
    Stop-safe admissibility (static obstacles):
      stopping distance d_stop = v^2 / (2*a_brake)
      require clearance > d_stop + robot_radius
    """
    if clearance <= 0.0:
        return False
    d_stop = (max(0.0, v) ** 2) / max(1e-9, 2.0 * cfg.a_brake)
    return clearance > (d_stop + cfg.robot_radius)


def normalize(vals: List[float]) -> List[float]:
    lo, hi = min(vals), max(vals)
    if abs(hi - lo) < 1e-12:
        return [0.0 for _ in vals]
    return [(v - lo) / (hi - lo) for v in vals]


def dwa_control(s: State, goal: Tuple[float, float], obstacles: np.ndarray, cfg: Config) -> Tuple[float, float, List[State]]:
    """
    Compute (v*, w*) maximizing weighted score over sampled (v,w) within dynamic window.
    Returns best velocities and the associated trajectory.
    """
    vL, vU, wL, wU = dynamic_window(s, cfg)

    v_samples = np.arange(vL, vU + 1e-9, cfg.v_res)
    w_samples = np.arange(wL, wU + 1e-9, cfg.w_res)

    cand = []
    for v in v_samples:
        for w in w_samples:
            traj = rollout(s, float(v), float(w), cfg)
            clear = min_clearance(traj, obstacles) - cfg.robot_radius
            if not admissible(float(v), clear, cfg):
                continue
            h = heading_score(traj, goal)
            sp = speed_score(float(v), cfg)
            cand.append((h, clear, sp, float(v), float(w), traj))

    if not cand:
        # fallback: stop
        return 0.0, 0.0, [s]

    hs = normalize([c[0] for c in cand])
    cs = normalize([c[1] for c in cand])
    vs = normalize([c[2] for c in cand])

    best_j = -1e18
    best = cand[0]
    for i, c in enumerate(cand):
        j = cfg.w_heading * hs[i] + cfg.w_clear * cs[i] + cfg.w_speed * vs[i]
        if j > best_j:
            best_j = j
            best = c

    _, _, _, v_best, w_best, traj_best = best
    return v_best, w_best, traj_best


def wrap_to_pi(a: float) -> float:
    """Map angle to (-pi, pi]."""
    return (a + math.pi) % (2.0 * math.pi) - math.pi


def demo():
    """
    Simple closed-loop demo with point obstacles.
    Requires: numpy (and optionally matplotlib for visualization).
    """
    cfg = Config()
    s = State(x=0.0, y=0.0, theta=0.0, v=0.0, w=0.0)
    goal = (8.0, 6.0)

    # Obstacles as points
    obstacles = []
    for x in np.linspace(2.0, 7.0, 15):
        obstacles.append((x, 3.0))
    for y in np.linspace(1.0, 5.0, 12):
        obstacles.append((4.0, y))
    obs = np.array(obstacles, dtype=float)

    path = [(s.x, s.y)]
    for _ in range(600):
        v, w, _ = dwa_control(s, goal, obs, cfg)
        s = motion_step(s, v, w, cfg.dt)
        path.append((s.x, s.y))
        if math.hypot(s.x - goal[0], s.y - goal[1]) < 0.3:
            break

    print("Final state:", s)
    print("Steps:", len(path))

    try:
        import matplotlib.pyplot as plt
        path = np.array(path)
        plt.figure()
        plt.plot(path[:, 0], path[:, 1], label="trajectory")
        plt.scatter(obs[:, 0], obs[:, 1], s=15, label="obstacles")
        plt.scatter([goal[0]], [goal[1]], s=50, marker="*", label="goal")
        plt.axis("equal")
        plt.grid(True)
        plt.legend()
        plt.show()
    except Exception as e:
        print("Matplotlib plot skipped:", e)


if __name__ == "__main__":
    demo()
